test-e2e/src/halo2/mod.rs

```diff
@@ -0,0 +1,218 @@
+
+#[derive(Clone, Debug)]
+pub struct SerializablePublicInputs(pub Vec<Fr>);
+
+pub(crate) struct FibonacciMoproCircuit;
+
+impl FibonacciMoproCircuit {
+    pub(crate) fn prove(
+        srs_key_path: &str,
+        proving_key_path: &str,
+        input: HashMap<String, Vec<String>>,
+    ) -> Result<mopro_ffi::GenerateProofResult, MoproError> {
+        let circuit_inputs = deserialize_circuit_inputs(input).map_err(|e| {
+            MoproError::Halo2Error(format!("Failed to deserialize circuit inputs: {}", e))
+        })?;
+
+        let srs = halo2_fibonacci::io::read_srs_path(Path::new(&srs_key_path));
+
+        let proving_key =
+            halo2_fibonacci::io::read_pk::<FibonacciCircuit<Fr>>(Path::new(&proving_key_path));
+
+        let (proof, inputs) = generate_halo2_proof(&srs, &proving_key, circuit_inputs)
+            .map_err(|e| MoproError::Halo2Error(format!("Failed to generate the proof: {}", e)))?;
+
+        let serialized_inputs = bincode::serialize(&inputs).map_err(|e| {
+            MoproError::Halo2Error(format!("Serialisation of Inputs failed: {}", e))
+        })?;
+
+        Ok(mopro_ffi::GenerateProofResult {
+            proof,
+            inputs: serialized_inputs,
+        })
+    }
+
+    pub(crate) fn verify(
+        srs_key_path: &str,
+        verifying_key_path: &str,
+        proof: Vec<u8>,
+        public_inputs: Vec<u8>,
+    ) -> Result<bool, MoproError> {
+        let deserialized_inputs: SerializablePublicInputs = bincode::deserialize(&public_inputs)
+            .map_err(|e| MoproError::Halo2Error(e.to_string()))?;
+
+        let srs = halo2_fibonacci::io::read_srs_path(Path::new(&srs_key_path));
+
+        let verifying_key =
+            halo2_fibonacci::io::read_vk::<FibonacciCircuit<Fr>>(Path::new(&verifying_key_path));
+
+        let is_valid =
+            verify_halo2_proof(&srs, &verifying_key, proof, deserialized_inputs).unwrap();
+
+        Ok(is_valid)
+    }
+}
+
+fn generate_halo2_proof(
+    srs: &ParamsKZG<Bn256>,
+    proving_key: &ProvingKey<G1Affine>,
+    inputs: CircuitInputs,
+) -> Result<(Halo2Proof, SerializablePublicInputs), MoproError> {
+    let circuit = FibonacciCircuit::<Fr>::default();
+
+    // Generate the proof - so far using dummy inputs, will be replaced with actual inputs
+
+    let a = Fr::from(1); // F[0]
+    let b = Fr::from(1); // F[1]
+
+    // `out` value right now must be 55, but will be replaced with the actual output value
+    let out: Fr = inputs
+        .get("out")
+        .ok_or(MoproError::Halo2Error(
+            "Failed to get `out` value".to_string(),
+        ))?
+        .get(0)
+        .ok_or(MoproError::Halo2Error(
+            "Failed to get `out` value".to_string(),
+        ))?
+        .clone();
+
+    let mut transcript = TranscriptWriterBuffer::<_, G1Affine, _>::init(Vec::new());
+
+    let public_input = vec![a, b, out];
+
+    create_proof::<KZGCommitmentScheme<Bn256>, ProverGWC<_>, _, _, Blake2bWrite<_, _, _>, _>(
```

Comment:
I think this logic should also be in `halo2_fibonacci`, specifically the generics about commitment scheme and field.

Reply:
Ideally the code here is just `halo2_fibonacci::prove(srs_path, proving_key, inputs)`